[Yap Yuhang] iP

yyhangz opened the pull request:
# CS2103T iP

Tired of writing down and organizing your own task lists? 🤔 

Fret not! Introducing Duke, a simple task organizer built using Java~~script~~. It
- is text-based
- has intuitive and simple commands
- is easy to learn

To use Duke,
1. Check out our [Github page](https://github.com/yyhangz/ip)
2. Follow our installation guides
3. Add tasks
4. Let Duke handle the rest!

Duke has interesting features such as:
- [X] Managing tasks
- [X] Filtering tasks
- [ ] Alerts and notifications (coming soon)

Here's an exclusive sneak peek of the `main` function to see how Duke was developed!
```java
public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }
```

And always remember,
> Duke recruits the best basketball players. They don't recruit dunkers or highlight makers. They recruit good basketball players. - Zion Williamson

Commit:
Add Gradle support

yyhangz committed:
Add echo funtionality

yyhangz committed:
Add task list functionality

yyhangz committed:
Add mark-as-done feature

yyhangz committed:
Add Todo, Event and Deadline task types

yyhangz committed:
Add IO redirection testing

yyhangz committed:
Add error handling for duke commands

yyhangz committed:
Add delete task functionality

yyhangz committed:
Reformat code based on coding standards

yyhangz committed:
Add functionality to cache and load past user inputs

yyhangz committed:
Reformat code to abide module coding standards

yyhangz committed:
Add java date format for Event and Deadline

yyhangz committed:
Add java date format for event and deadline task types

yyhangz committed:
Refactor code to extract out closely related code as classes

yyhangz committed:
Refactor code to integrate packaging system

yyhangz committed:
Add Junit tests for Todo and DukeParser classes

yyhangz committed:
Add manifest for Duke jar file for initial release

yyhangz committed:
Update JavaDoc for public classes and methods

yyhangz committed:
Update coding style to conform to module style guide

yyhangz committed:
Add find-task-by-keyword feature

yyhangz committed:
Merge branch 'branch-A-JavaDoc'

yyhangz committed:
Merge branch 'branch-A-CodingStandard'

yyhangz committed:
Add find-tasks-by-keyword feature

SHEZADHASSAN22 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+     * Returns a string that represents a serialized store format of the task that is specific to Duke.
+     * @return a string of serialized format.
+     */
+    @Override
```
There should be an empty line between description and parameter section of the javadoc comment.

SHEZADHASSAN22 commented on src/main/java/duke/DukeParser.java:
```diff
@@ -0,0 +1,104 @@
+            task = new Event(taskDescription, date);
+            break;
+        default:
+            throw new DukeFileSystemException("Invalid data read from file while loading tasks. A new list will " +
```
The plus operator should go on the second line.

SHEZADHASSAN22 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,52 @@
+     */
+    @Override
+    public String toString() {
+        return String.format("[E] %s (at: %s)", super.toString(), this.eventTime.format(DateTimeFormatter
```
I think it would read better if you started the second line after the last comma.

SHEZADHASSAN22 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Constructs a new instance of Storage that loads and stores data into the file path that is specified.
```
There should be a gap between the description and parameters in the javadoc comment.

yyhangz replied:
Thanks for spotting these mistakes! I will make the necessary changes :)

yyhangz committed:
Merge branch 'add-gradle-support'

yyhangz committed:
Fix mistakes in code style

yyhangz committed:
Integrate gradle into workflow

yyhangz committed:
Merge branch 'A-gradle'

yyhangz committed:
Edit Event.java to match module style guide

yyhangz committed:
Merge branch 'branch-A-CheckStyle'

yyhangz committed:
Add GUI interface for Duke

yyhangz committed:
Merge branch 'branch-Level-10'

yyhangz committed:
Add GUI interface for Duke

yyhangz committed:
Fix Deadline bug

yyhangz committed:
Refactor duke packages

yyhangz committed:
Remove redundant Ui class

houtenteo commented on src/main/java/duke/io/DukeParser.java:
```diff
@@ -0,0 +1,109 @@
+     * Constructs a new instance of the DukeParser. No arguments are expected.
+     */
+    public DukeParser() {
+
```
I think omitting the constructor would be better for readability. Leaving it blank could lead to confusion.

houtenteo commented on src/main/java/duke/io/Ui.java:
```diff
@@ -0,0 +1,46 @@
+     * Constructs a new instance of Ui. No arguments are expected.
+     */
+    public Ui() {
+
```
Likewise for this blank constructor as well.

houtenteo commented on src/main/java/duke/io/DukeParser.java:
```diff
@@ -0,0 +1,109 @@
+     * @return a corresponding Duke command type based on the input string.
+     */
+    public DukeCommand getCommandType(String inputString) {
+        DukeCommand inputCommand;
```
Even for a long method like this, you still managed to avoid the arrowhead style. Made it way easier to read and understand. Good job!

houtenteo commented on src/main/java/duke/io/Ui.java:
```diff
@@ -0,0 +1,46 @@
+     * @param s The error message to be printed.
+     */
+    public void dukeShowError(String s) {
+        Exception e = new InvalidDukeCommandException(s);
```
Method names should be verbs. Would be better to name this method 'showDukeError' instead.

houtenteo commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,271 @@
+            return "The description of an event cannot be empty.";
+        }
+        String taskDescription = args.split(" /at ")[0];
+        String eventTime;
```
Since this eventTime variable is only used within the try block, it is better to declare it within the try block itself. Check out the [textbook](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#minimize-scope-of-variables) on minimizing scope of variables.

yyhangz replied:
Thanks for the comments!

houtenteo reviewed (commented):
Very easy to read and understand. Great work overall.

yyhangz committed:
Reformat code to adhere to module coding standards

yyhangz committed:
Merge branch 'branch-refactor'

yyhangz committed:
Update grammar in Duke return message for list function

yyhangz committed:
Add assertions to check for null values in critical functions

These assertions reduce the possibility of a major error
occurring in the actual application by asserting that critical
variables are not defined as null.

This halts the problem from escalating to the user before
it is too late.

yyhangz committed:
Edit code base to adhere to module coding quality

Fix several minor issues like choice of variable names
and line breaks in JavaDocs.

Also made several major changes such as redefining certain
constants as class-level constants to avoid the use of
magic literals.

yyhangz committed:
Merge pull request #1 from yyhangz/branch-A-Assertions

Add assertions to check for null values in critical functions

yyhangz committed:
Merge branch 'master' of https://github.com/yyhangz/ip

yyhangz committed:
Merge branch 'master' into branch-A-CodeQuality

yyhangz committed:
Merge pull request #2 from yyhangz/branch-A-CodeQuality

Edit code base to adhere to module coding quality

yyhangz committed:
Merge branch 'master' of https://github.com/yyhangz/ip

yyhangz committed:
Update gitignore to remove data file

yyhangz committed:
Refactor code to adhere to module code style guide

yyhangz committed:
Remove height limit for Duke dialog box

yyhangz committed:
Integrate ability to do done and delete operations on multiple tasks at once

yyhangz committed:
Merge branch 'branch-C-MassOps'

yyhangz committed:
Reformat test units to adhere to checkstyle

yyhangz committed:
Create user guide

yyhangz committed:
Set theme jekyll-theme-cayman

yyhangz committed:
Set theme jekyll-theme-slate

yyhangz committed:
Add feature to close GUI with bye command

yyhangz committed:
Merge branch 'master' of https://github.com/yyhangz/ip
